[OEP 4071]: restore to a new volume from LVM thin snapshot (#4080)
openebs/openebs: designs/local-pv/lvm/snapshot-restore.md

---
oep-number: OEP 4071
title: Add Snapshot Restore Feature to OpenEBS LVM
authors:
  - "@rohan2794"
owners:
  - "@rohan2794"
editor: TBD
creation-date: 03/10/2025
last-updated: 03/10/2025
status: provisional
---

# Add Snapshot Restore Feature to OpenEBS LVM
> Warning: Restoring from a snapshot is available exclusively for thinly provisioned volume snapshots.

## Table of Contents

1. [Overview](#overview)
2. [Motivation](#motivation)
3. [Goals](#goals)
4. [Non-Goals](#non-goals)
5. [Proposal](#proposal)
6. [User Stories](#user-stories)
7. [Implementation Details](#implementation-details)
8. [Testing](#testing)

## Overview

This OEP proposes adding Snapshot Restore feature to OpenEBS LVM.

> This feature would enable users to create restore from snapshot of thinly provisioned volume.

Reviewer: ```suggestion
This feature would enable users to create restore from thin snapshot LV (LVM Logical Volume).
```

Reviewer: perhaps you may define a terminology section to e.g indicate `LV` is `LVM Logical Volume` , or `VG` is `LVM Volume Group` etc.

Author: add Terminology section

## Motivation

OpenEBS LVM supports snapshots, but restoring from them is not supported. This limitation creates challenges when attempting to recover data from snapshots.

## Goals

- Enable users to create restore from snapshot of thinly provisioned volume.

## Non-Goals

> - Restoring from snapshots of thick-provisioned volumes.

Reviewer: ```suggestion
- Restoring from thick snapshot LVs.
```

## Proposal

> For restoring from snapshot of thin-provisioned volume, a new thin-provisioned snapshot is created from the existing snapshot. Since LVM snapshots are writable, these can be exposed as new PVCs without losing the original snapshot’s identity. This allows multiple restores from the same snapshot, resulting in multiple independent copies. The restored thin snapshot volume is activated as Read-Write and presented to Kubernetes as PersistentVolume object.

Reviewer: ```suggestion
For restoring from snapshot of thin-provisioned volume, a new thin snapshot LV is created from an existing thin snapshot LV. Since LVM snapshots are writable, these can be exposed as new PVCs without losing the original snapshot’s identity. This allows multiple restores from the same snapshot, resulting in multiple independent copies. The restored thin snapshot volume is activated as Read-Write and presented to Kubernetes as PersistentVolume object.
```

- Add two new fields to the LVM Snapshot CR: **`thinSnapshot`** and **`sourceVolumeCapacity`**.
    - **`thinSnapshot`**: Indicates whether the snapshot was created from a thin-provisioned volume.  
    - **`sourceVolumeCapacity`**: Records the capacity of the origin volume at the time the snapshot was taken.
- Add a new field **`source`** to the LVM Volume CR. It defines the data source for the volume.
    - Can be a **snapshot**  
    - Or an **existing volume**

## User Stories

### Story 1

As a user, I want to create a `PersistentVolumeClaim` from a `VolumeSnapshot` of a thin-provisioned volume.

> 

Reviewer: ```suggestion
```

## Implementation Details

### Design

- With LVM Thin Pools, snapshot of snapshot can be writable and used as lightweight clones.
- Common blocks in nested snapshots are stored once in the thin pool, so snapshot chains can grow without limits or performance loss.

> - To restore a PVC from LVM snapshot of thin-provisoned volume, a new thin-provisioned snapshot is created from the existing snapshot. Since LVM snapshots are inherently writable, this snapshot can be exposed to the user as a volume for a new PVC.

Reviewer: It may be worth mentioning somewhere that even though they are default writable, the normal snapshots that openebs lvm provisioner takes today are read-only. Only the ones taken as part of restore are made writable.

Author: added Note in the document

- The identity of the original LVM snapshot remains intact.

> - Restored thin snapshot volumes are activated in Read-Write mode and presented to Kubernetes users as PersistentVolume objects.

Reviewer: ```suggestion
- Restored thin snapshot volumes are activated in Read-Write mode and presented to applications as PersistentVolume objects.
```

- Writes to restored LV consume new blocks and Original snapshot remains point-in-time and read-only from Kubernetes perspective. This ensures multiple independent PVCs can be provisioned from a single snapshot safely.

### Workflow

1. User create PVC with data source as thin snapshot.
2. The **external-provisioner** kubernetes sidecar sends a CreateVolume gRPC call to OpenEBS LVM CSI driver.
3. This **CreateVolume** request containing a dataSource is received by the lvm-controller.
4. From the **CreateVolume** request, validate volume creation request
    - If **req.VolumeContentSource** is nil, proceed as normal OpenEBS LVM volume creation.

>     - On the other hand, the Volume creation request might have a data source as either a Snapshot or Persistent Volume.

Reviewer: if the source is another volume (PVC/PV), then return error?

Author: yes, i will update the doc

    - Validate request whether snapshot is thin or thick , if snapshot is thick then return not supported error.
    - Validate LVM vg , owner node and capacity of CreateVolume request.
    - If data source is snapshot then create volume with datasource as snapshot.
5. OpenEBS LVM controller creates lvm volume CR with snapshot name, node name, vgpattern,volGroup, capacity, thinProvision values.
6. **lvm-node** on the target node finds the earlier created restore volume lvmvolume CR with snapshot as source.

> 7. **lvm-node** sends a volume create request to create a thin-snapshot from the thin-snapshot

Reviewer: ```suggestion
7. **lvm-node** sends a volume create request to create a new thin snapshot LV from the source thin snapshot LV
```

spec:
  storageClassName: restore-sc
  dataSource:
    name: lvm-thin-snapshot # name of the thin snapshot
    kind: VolumeSnapshot
    apiGroup: snapshot.storage.k8s.io
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 2Gi # capacity of restored PVC
```

### OpeneEBS lvmvolume CR for restore

```yaml
apiVersion: local.openebs.io/v1alpha1
kind: LVMVolume
metadata:
  name: pvc-97e44bbf-e916-4fe4-bca0-04e5aabcba4e
  namespace: openebs
spec:
  capacity: "2147483648"
  ownerNodeID: node-0-397958
  source: snapshot-0e1967eb-471d-4a7e-93b1-ba92d53c7a2a # lvm snapshot CR name
  volGroup: lvmvg
status:
  state: Ready
```

### Risks and Mitigations

- Enable **thin_pool_autoextend_threshold** and **thin_pool_autoextend_percent** in **lvm.conf**. Prevents write failures when thin pool fills. Ensure Volume Group (VG) has spare capacity, otherwise auto-extend won’t help.

## Testing

- Restore from a thin snapshot.  
- Restore from a thick snapshot — operation should fail.  
- Restore using a non-existent snapshot — operation should fail.  
- Restore from a snapshot belonging to a different volume group (VG) — operation should fail.  

> - Expand a restored volume created from a thin snapshot.  

Reviewer: should succeed?

- Attempt to create a restored PVC with capacity greater than the source volume at snapshot time — operation should fail.  
- Attempt to create a restored PVC with capacity smaller than the source volume at snapshot time — operation should fail.  

> - Delete the snapshot after performing a restore.

Reviewer: should succeed?

---
oep-number: OEP 4071
title: Add Snapshot Restore Feature to OpenEBS LVM
authors:
  - "@rohan2794"
owners:
  - "@rohan2794"
editor: TBD
creation-date: 03/10/2025
last-updated: 09/10/2025

> status: provisional

Reviewer 2: This feature had already been added to the roadmap [here](https://github.com/openebs/openebs/blob/17c908417d27be5c628a9003a17dcbcc4fd057b0/ROADMAP.md?plain=1#L32) so I think you can go ahead and mark this as implementable

Author: ok

> Note: By default, LVM snapshots are writable. However, the snapshots currently created by the OpenEBS LVM provisioner are read-only. Only snapshots generated during restore operations by the LVM backend are made writable.

### Workflow

1. User create PVC with data source as thin snapshot.
2. The **external-provisioner** kubernetes sidecar sends a CreateVolume gRPC call to OpenEBS LVM CSI driver.
3. This **CreateVolume** request containing a dataSource is received by the lvm-controller.
4. From the **CreateVolume** request, validate volume creation request
    - If **req.VolumeContentSource** is nil, proceed as normal OpenEBS LVM volume creation.
    - On the other hand, the Volume creation request might have a data source as either a Snapshot or Persistent Volume.
    - Validate request whether snapshot is thin or thick , if snapshot is thick then return not supported error.
    - Validate LVM vg and capacity of CreateVolume request.
    - If the source is another volume (PVC/PV), then return not supported error.
    - If data source is snapshot then create volume with datasource as snapshot.

> 5. OpenEBS LVM controller creates lvm volume CR with snapshot name, node name, vgpattern,volGroup, capacity, thinProvision values.

Reviewer 3: ```suggestion
5. OpenEBS LVM controller creates lvm volume CR with snapshot name, node name, vgpattern, volGroup, capacity, thinProvision values.
```

6. **lvm-node** on the target node finds the earlier created restore volume lvmvolume CR with snapshot as source.

> 7. **lvm-node** sends a volume create request to create a new thin snapshot LV from the source thin snapshot LV

Reviewer 2: This is done locally right?

```suggestion
7. **lvm-node** issues a volume create request to create a new thin snapshot LV from the source thin snapshot LV
```

    - Backend (LVM) could internally:
        - Creates thin snapshot of snapshot
            - ```lvcreate -s -n openebs-lvm-restore-lv lvmvg/openebs-lvm-thin-snapshot```
        - Activates above snapshot with Read-Write permissions which will be restored LV
            - ```lvchange -kn -ay lvmvg/openebs-lvm-restore-lv```
8. **lvm-node** updates the status of restored LVM Volume CR
9. **lvm-controller** finds the updated status of restored LVM Volume CR

> 10. **lvm-controller** sends the success (or failure) to provisioner sidecar.

Reviewer 2: There's no path from controller to side car, it's part of the create volume issues from side car to the controller

```suggestion
10. **lvm-controller** complete create_volume request with success (or failure).
```

    else req.VolumeContentSource is Snapshot or PV
        Controller->>Controller: Validate data source type
        alt req.VolumeContentSource is PV
            Controller-->>Provisioner: Return "not supported" error
        else req.VolumeContentSource is Snapshot
            Controller->>Controller: Validate snapshot type
            alt Snapshot is thick
                Controller-->>Provisioner: Return "not supported" error
            else Snapshot is thin
                Controller->>Controller: Validate LVM vg and capacity
                Controller->>Controller: Create volume with snapshot datasource
            end
        end
    end
    Controller-->>K8sAPI: Create LVM Volume CR
    Node->>Node: Detects LVM Volume CR <br>with snapshot as source
    Node->>Backend: Create new LVM LV <br>from data source thin snapshot
    Backend->>Backend: create thin snapshot <br>from thin snapshot
    Backend->>Backend: activate thin snapshot
    Backend-->>Node: Report LVM LV creation status
    Node->>Node: Update status of restored <br>LVM Volume CR
    Controller->>Controller: Find updated status of LVM Volume CR

>     Controller-->>Provisioner: Send success/failure response

Reviewer 2: ```suggestion
    Controller-->>Provisioner: Complete Request with success/failure response
```

    Provisioner-->>K8sAPI: Forward response
    K8sAPI-->>User: Return result
```

<br>

### OpenEBS lvmsnapshot CR

```yaml
apiVersion: local.openebs.io/v1alpha1
kind: LVMSnapshot
metadata:
  name: snapshot-0e1967eb-471d-4a7e-93b1-ba92d53c7a2a # lvm snapshot cr name
  namespace: openebs
spec:
  ownerNodeID: node-0-397958
  # origin volume capacity at the time of snapshot

>   sourceVolumeSize: "2147483648"

Reviewer 4: sourceVolumeSize can be different when volumes is resized after taking snapshot. I think we do not allow resize after taking snapshot. Maybe we can allow that in future as we verified IO on extended region doesn't cause any failure.

sourceVolumeSize doesn't really imply it was size of volume while taking snapshot. Shall we just call it `size` ?

Author: I have updated it to `snapOriginVolumeCapacity` .

Reviewer 4: Its actually the snapshot size itself right? `snapOriginVolumeCapacity` still seems like the original volume size which could have got resized? just `size` seems simpler.

Author: No , it's not snapshot size, it's size of origin volume at the time of snapshot.

  # 'thinSnapshot' specifies type of snapshot, this field will be yes 
  # if snapshot is of thin-provisoned volume else it will be no

>   thinSnapshot: "yes"

Reviewer 4: Lets just call it `thin`. Its part of lvmsnapshot CR anyways. Also, Can we make it bool?

Author: ok. 
LVM volume CR also have thin value as yes or no, so i have kept it same  as LVM volume CR.

Reviewer 4: That was not correct i feel. Moving forward sticking to bool way would be better.

Author: make sense, i will change it to bool

  volGroup: lvmvg
status:
  state: Ready
```

### Restore from Snapshot

***restore-pvc.yaml***
```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: pvs-restore-from-snapshot
spec:
  storageClassName: restore-sc
  dataSource:
    name: lvm-thin-snapshot # name of the thin snapshot
    kind: VolumeSnapshot
    apiGroup: snapshot.storage.k8s.io
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 2Gi # capacity of restored PVC
```

### OpeneEBS lvmvolume CR for restore

```yaml
apiVersion: local.openebs.io/v1alpha1
kind: LVMVolume
metadata:
  name: pvc-97e44bbf-e916-4fe4-bca0-04e5aabcba4e
  namespace: openebs
spec:
  capacity: "2147483648"
  ownerNodeID: node-0-397958

>   source: snapshot-0e1967eb-471d-4a7e-93b1-ba92d53c7a2a # lvm snapshot CR name

Reviewer 4: This will be empty for a `lvmvolume` without source?

Author: yes it will be empty when there is no source

  volGroup: lvmvg
status:
  state: Ready
```

### Risks and Mitigations

- Enable **thin_pool_autoextend_threshold** and **thin_pool_autoextend_percent** in **lvm.conf**. Prevents write failures when thin pool fills. Ensure Volume Group (VG) has spare capacity, otherwise auto-extend won’t help.

## Testing

> - Restore from a thin snapshot.  
> - Restore from a thick snapshot — operation should fail.  

Reviewer 2: repeated

```suggestion
- Restore from a thick snapshot — operation should fail.  
```

- Restore using a non-existent snapshot — operation should fail.  
- Restore from a snapshot belonging to a different volume group (VG) — operation should fail.  
- Expansion of restored volume created from a thin snapshot should succeed.  

> - Attempt to create a restored PVC with capacity greater than the source volume at snapshot time — operation should fail.  

Reviewer 2: I was wondering if we could support this by resizing the filesystem at create time?

Author: yes, it can be achieved by resizing filesystem during restore later

## Implementation Details

### Design

- With LVM Thin Pools, snapshot of snapshot can be writable and used as lightweight clones.
- Common blocks in nested snapshots are stored once in the thin pool, so snapshot chains can grow without limits or performance loss.
- To restore a PVC from LVM snapshot of thin-provisoned volume, a new thin-provisioned snapshot is created from the existing snapshot. Since LVM snapshots are inherently writable, this snapshot can be exposed to the user as a volume for a new PVC.
- The identity of the original LVM snapshot remains intact.
- Restored thin snapshot volumes are activated in Read-Write mode and presented to applications as PersistentVolume objects.
- Writes to restored LV consume new blocks and Original snapshot remains point-in-time and read-only from Kubernetes perspective. This ensures multiple independent PVCs can be provisioned from a single snapshot safely.

> Note: By default, LVM snapshots are writable. However, the snapshots currently created by the OpenEBS LVM provisioner are read-only. Only snapshots generated during restore operations by the LVM backend are made writable.

### Workflow

> 1. User create PVC with data source as thin snapshot.

Reviewer: ```suggestion
1. User creates PVC with data source as thin snapshot.
```
